test/test_calculate_cca.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import division
+
+import unittest
+
+import numpy as np
+np.random.seed(42)
+
+from wyrm.types import Data
+from wyrm.processing import append, swapaxes, calculate_cca
+
+
+class TestCalculateCCA(unittest.TestCase):
+
+    SAMPLES = 100
+    CHANNELS_X = 10
+    CHANNELS_Y = 5
+
+    def setUp(self):
+        # create a random noise signals with
+        # X: 100 samplse and 10 channels
+        # Y: 100 samples and 5 channels
+        self.X = np.random.randn(self.SAMPLES, self.CHANNELS_X)
+        self.Y = np.random.randn(self.SAMPLES, self.CHANNELS_Y)
+        axes_x = [np.arange(self.X.shape[0]), np.arange(self.X.shape[1])]
+        axes_y = [np.arange(self.Y.shape[0]), np.arange(self.Y.shape[1])]
+        self.dat_x = Data(self.X, axes=axes_x, names=['time', 'channel'], units=['ms', '#'])
+        self.dat_y = Data(self.Y, axes=axes_y, names=['time', 'channel'], units=['ms', '#'])
+
+    def test_rho(self):
+        """Test if the canonical correlation is between 0 and 1."""
+        rho, w_x, w_y = calculate_cca(self.dat_x, self.dat_y)
+        self.assertTrue(0 < rho <1)
+
+    def test_raise_error_with_non_continuous_data(self):
+        """Raise error if ``dat_x`` is not continuous Data object."""
+        dat = Data(np.random.randn(2, self.SAMPLES, self.CHANNELS_X),
+                   axes=[[0, 1], self.dat_x.axes[0], self.dat_x.axes[1]],
+                   names=['class', 'time', 'channel'],
+                   units=['#', 'ms', '#'])
+        with self.assertRaises(AssertionError):
+            calculate_cca(dat, self.dat_x)
+
+    def test_raise_error_with_different_length_data(self):
+        """Raise error if the length of ``dat_x`` and ``dat_y`` is different."""
+        dat = append(self.dat_x, self.dat_x)
+        with self.assertRaises(AssertionError):
+            calculate_cca(dat, self.dat_y)
+
+    def test_calculate_csp_swapaxes(self):
+        """caluclate_cca must work with nonstandard timeaxis."""
+        res1 = calculate_cca(swapaxes(self.dat_x, 0, 1), swapaxes(self.dat_y, 0, 1), timeaxis=1)
+        res2 = calculate_cca(self.dat_x, self.dat_y)
+        np.testing.assert_array_equal(res1[0], res2[0])
+        np.testing.assert_array_equal(res1[1], res2[1])
+        np.testing.assert_array_equal(res1[2], res2[2])
+
+    def test_calculate_cca_copy(self):
+        """caluclate_cca must not modify argument."""
+        cpy_x = self.dat_x.copy()
+        cpy_y = self.dat_y.copy()
+        calculate_cca(self.dat_x, self.dat_y)
+        self.assertEqual(self.dat_x, cpy_x)
+        self.assertEqual(self.dat_y, cpy_y)
+
```

Comment:
Hey Keita,

in prinicple CCA doesn't care about the temporal structure of the data.
I.e. it will work with random data just as fine as it will with
oscillations, as long as there really is a hidden correlation in the data.

I have prepared a little Matlab script that you can use as the basis for a
test. The first example in the script is similar to the case you described,
where you have one multivariate target signal (EEG data for example) as X
and then for Y the sine and cosine of the target frequency. The second case
in the script is the more general one, where the target signal is
completely random but it is mixed into both datasets X and Y. In both
cases, CCA should be able to extract the target signal from X and Y.

Here comes the Matlab code. If you don't have Matlab, it should be straight
forward to transform it into python code.
Cheers,
Sven

PS: the obtained correlations depend on the noise_level


%% parameters

CHANNELS_X = 15;
CHANNELS_Y = 17;
SAMPLES = 1000;
FREQUENCY = 10;

t = linspace(0,1,SAMPLES)';
noise_level = 0.1;

%% example 1 --> steady state BCI

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);
Sx(:,1) = sign(randn(1,1)) * sin(2 * pi * FREQUENCY * t + randn() * pi) +
noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% here only sin and cos of same frequency
Sy = [sin(2 * pi * FREQUENCY * t), cos(2 * pi * FREQUENCY * t)];
Y = Sy;

% perform CCA
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])

fprintf('r = %g\n',r(1))


%% example 2 --> general case

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% source signals for Y
Sy = randn(SAMPLES, CHANNELS_Y);
% make the first source signal the same as in the X sources, perhaps plus
% some small amount of noise
Sy(:,1) = Sx(:,1) + noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ay = randn(CHANNELS_Y,CHANNELS_Y);
Y = Sy * Ay;

% perform
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])

fprintf('r = %g\n',r(1))







2015-01-07 14:46 GMT+01:00 Keita Kagurazaka <notifications@github.com>:

> In test/test_calculate_cca.py
> <https://github.com/venthur/wyrm/pull/1#discussion-diff-22586094>:
>
> > +    def test_calculate_csp_swapaxes(self):
> > +        """caluclate_cca must work with nonstandard timeaxis."""
> > +        res1 = calculate_cca(swapaxes(self.dat_x, 0, 1), swapaxes(self.dat_y, 0, 1), timeaxis=1)
> > +        res2 = calculate_cca(self.dat_x, self.dat_y)
> > +        np.testing.assert_array_equal(res1[0], res2[0])
> > +        np.testing.assert_array_equal(res1[1], res2[1])
> > +        np.testing.assert_array_equal(res1[2], res2[2])
> > +
> > +    def test_calculate_cca_copy(self):
> > +        """caluclate_cca must not modify argument."""
> > +        cpy_x = self.dat_x.copy()
> > +        cpy_y = self.dat_y.copy()
> > +        calculate_cca(self.dat_x, self.dat_y)
> > +        self.assertEqual(self.dat_x, cpy_x)
> > +        self.assertEqual(self.dat_y, cpy_y)
> > +
>
> I think CCA cannot recovery source components from random mixed signals.
>
> For testing the functionality of my CCA implementation, I considered a
> following case.
>
>    - X is a data set which includes a signal called S.
>    - Y is a data set which includes orthonormal basis whose linear span
>    includes the S.
>
> In this case, the first canonical correlation coefficient almost equals 1
> and the canonical variable of Y almost equals S after zero-meaning and
> scaling.
>
> Pseudo-code is here.
>
> X = randn(SAMPLES, CHANNELS_X)
> Y = randn(SAMPLES, CHANNELS_Y)
> X[:, 0] = randn() * sin(2 * pi * FREQUENCY * t + randn() * pi) + randn()
> Y[:, 0] = sin(2 * pi * FREQUENCY * t)
> Y[:, 1] = cos(2 * pi * FREQUENCY * t)
>
> rho, w_x, w_y = calculate_cca(X, Y)# rho almost equals 1.0# dot(Y, w_y) almost equals X[:, 0] after zero-meaning and scaling.
>
> What do you think about this test?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/venthur/wyrm/pull/1/files#r22586094>.
>